### PR TITLE
feat(sdk): Add support for state events in the timeline

### PR DESCRIPTION
**Pull request by zecakeh**
Tested locally in Fractal.

Signed-off-by: Kévin Commaille <zecakeh@tedomum.fr>

### CODE REVIEW

**Comment by zecakeh** [comment]
Ping @jplatte again for review.

File: crates/matrix-sdk/src/room/timeline/event_handler.rs [hunk 1]
```diff
@@ -280,9 +306,11 @@ impl<'a, 'i> TimelineEventHandler<'a, 'i> {
...
                 self.handle_redaction(redacts, content);
             }
-            TimelineEventKind::State { .. } | TimelineEventKind::RedactedState => {
-                // TODO
-                debug!("Ignoring state event, not supported yet");
+            TimelineEventKind::RoomMember { user_id, content, sender } => {
+                self.add(NewEventTimelineItem::room_member(user_id, content, sender))
+            }
+            TimelineEventKind::OtherState { state_key, content } => {
+                self.add(NewEventTimelineItem::other_state(state_key, content))
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_handler.rs:313]:
```suggestion
                self.add(NewEventTimelineItem::room_member(user_id, content, sender));
            }
            TimelineEventKind::OtherState { state_key, content } => {
                self.add(NewEventTimelineItem::other_state(state_key, content));
```

File: crates/matrix-sdk/src/room/timeline/event_item.rs [hunk 2]
```diff
@@ -451,3 +479,198 @@ impl Sticker {
...
+
+    /// The full content of the event.
+    pub fn content(&self) -> &FullStateEventContent<RoomMemberEventContent> {
+        &self.content
+    }
+
+    /// The membership change induced by this event.
+    ///
+    /// If this returns `None`, it doesn't mean that there was no change, but
+    /// that the change could not be computed. This is always the case with
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_item.rs:505]:
```suggestion
    /// that the change could not be computed. This is currently always the case with
```

File: crates/matrix-sdk/src/room/timeline/event_item.rs [hunk 3]
```diff
@@ -451,3 +479,198 @@ impl Sticker {
...
+            AnyFullStateEventContent::RoomName(c) => Self::RoomName(c),
+            AnyFullStateEventContent::RoomPinnedEvents(c) => Self::RoomPinnedEvents(c),
+            AnyFullStateEventContent::RoomPowerLevels(c) => Self::RoomPowerLevels(c),
+            AnyFullStateEventContent::RoomServerAcl(c) => Self::RoomServerAcl(c),
+            AnyFullStateEventContent::RoomThirdPartyInvite(c) => Self::RoomThirdPartyInvite(c),
+            AnyFullStateEventContent::RoomTombstone(c) => Self::RoomTombstone(c),
+            AnyFullStateEventContent::RoomTopic(c) => Self::RoomTopic(c),
+            AnyFullStateEventContent::SpaceChild(c) => Self::SpaceChild(c),
+            AnyFullStateEventContent::SpaceParent(c) => Self::SpaceParent(c),
+            AnyFullStateEventContent::_Custom { redacted, .. } => {
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_item.rs:621]: We should not be relying on `_Custom`. Maybe we can add an `is_redacted` accessor in Ruma (if we need that field) and make this the `_ =>` branch?

File: crates/matrix-sdk/src/room/timeline/event_item.rs [hunk 4]
```diff
@@ -451,3 +479,198 @@ impl Sticker {
...
+                Self::_Custom { event_type: event_type.to_string(), redacted }
+            }
+            AnyFullStateEventContent::RoomMember(_) => unimplemented!(),
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_item.rs:624]:
```suggestion
            AnyFullStateEventContent::RoomMember(_) => unreachable!(),
```

File: crates/matrix-sdk/src/room/timeline/event_item.rs [hunk 5]
```diff
@@ -451,3 +479,198 @@ impl Sticker {
...
+            // TODO: use `non_exhaustive_omitted_patterns` when it is stabilized.
+            // See <https://github.com/rust-lang/rust/issues/89554>
+            _ => unimplemented!(),
+        }
+    }
+
+    /// Get the event's type, like `m.room.create`.
+    pub fn event_type(&self) -> StateEventType {
+        match self {
+            AnyOtherFullStateEventContent::PolicyRuleRoom(c) => c.event_type(),
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/event_item.rs:634]:
```suggestion
            Self::PolicyRuleRoom(c) => c.event_type(),
```

File: crates/matrix-sdk/src/room/timeline/tests.rs [hunk 6]
```diff
@@ -542,6 +544,178 @@ async fn initial_events() {
     assert_eq!(items[2].as_event().unwrap().sender(), *BOB);
 }
 
+#[async_test]
+async fn other_state() {
+    let timeline = TestTimeline::new(&ALICE);
+    let mut stream = timeline.stream();
+
+    timeline
+        .handle_live_custom_event(json!({
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/tests.rs:553]: Instead of `_custom_event`, can you add `handle_live_state_event` (for `EmptyStateKey`) and `handle_live_state_event_with_state_key`, akin to `handle_live_message_event` and use that?

File: crates/matrix-sdk/src/room/timeline/event_item.rs (hunk 3 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/event_item.rs:621]: Ah right this branch is there because the `non_exhaustive_omitted_patterns` lint wanted it when I tried it. But maybe they'll change the behavior before it's stabilized so I'll remove it.

**Comment by zecakeh** [comment]
Sorry about the force pushes, I'm trying to get the uniffi bindings test to pass.

File: crates/matrix-sdk/src/room/timeline/event_item.rs (hunk 3 above)
Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/event_item.rs:621]: No, I'm pretty sure we just can't use that lint here.

Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/event_item.rs]: I removed `is_redacted` because I doubt it'll be useful.

File: crates/matrix-sdk/src/room/timeline/tests.rs [hunk 7]
```diff
@@ -794,6 +786,60 @@ fn make_message_event<C: MessageLikeEventContent>(sender: &UserId, content: C) -
...
+        "origin_server_ts": next_server_ts(),
+        "unsigned": make_redacted_unsigned(sender, event_id),
+    })
+}
+
+fn make_redacted_unsigned(sender: &UserId, redacts: &EventId) -> JsonValue {
+    json!({
+        "redacted_because": {
+            "content": {},
+            "redacts": redacts,
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/tests.rs:834]: Because the `redacted_because` event can also be itself redacted and we would never read this field anyways, I think we can remove this here.

**Review by jplatte — APPROVED** [review]

File: bindings/matrix-sdk-ffi/src/timeline.rs [hunk 8]
```diff
@@ -614,3 +625,169 @@ pub impl MediaSourceExt for MediaSource {
         }
     }
 }
+
+#[derive(Clone, uniffi::Enum)]
+pub enum MembershipChange {
+    None,
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/timeline.rs:631]: Why do we have this variant? If nothing changed, should we really add a timeline item?

File: bindings/matrix-sdk-ffi/src/timeline.rs [hunk 9]
```diff
@@ -270,6 +279,8 @@ pub enum TimelineItemContentKind {
     RedactedMessage,
     Sticker { body: String, info: ImageInfo, url: String },
     UnableToDecrypt { msg: EncryptedMessage },
+    RoomMembership { change: MembershipChange },
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/timeline.rs:282]: Semi-related, is this for all `m.room.member` events? I'm thinking separate `MembershipChange` and `ProfileChange` variants might make sense (the latter would be a separate PR).

File: bindings/matrix-sdk-ffi/src/timeline.rs (hunk 8 above)
Comment (zecakeh) [reply bindings/matrix-sdk-ffi/src/timeline.rs:631]: Why would we arbitrarily remove an item from the timeline? Isn't it the client's job to ignore events they don't want to display?

Besides the item might contain data that changed that we don't know about (eg an extra profile field) in the raw JSON, but the client does.

File: bindings/matrix-sdk-ffi/src/timeline.rs (hunk 9 above)
Comment (zecakeh) [reply bindings/matrix-sdk-ffi/src/timeline.rs:282]: Sure we can do that. You put this comment in the FFI crate, but I'm guessing this is valid in the main crate too?

Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/timeline.rs:282]: Oh, yes 😅

**Comment by zecakeh** [comment]
Rebased on `main`.
